File: 1146-greatest-common-divisor-of-strings/greatest-common-divisor-of-strings.ex
defmodule Solution do
  @spec gcd_of_strings(str1 :: String.t, str2 :: String.t) :: String.t
  def gcd(a, b) when b != 0 do
    IO.puts("#{a}, #{b}")
    gcd(b, rem(a, b))
  end

  def gcd(a, _b), do: a

  def gcd_of_strings(str1, str2) do
    if str1 <> str2 != str2 <> str1 do
      ""
    else
      str1 |> String.slice(0, gcd(String.length(str1), String.length(str2)))
    end
  end
end

File: 1146-greatest-common-divisor-of-strings/greatest-common-divisor-of-strings.go
// Solution 2:
// Using S1+S2=S2+S1, Commutative property for strings!
func gcd(a, b int) int {
    for b != 0 {
        a, b = b, a%b
    }
    return a
}

func gcdOfStrings(str1 string, str2 string) string {
    if str1 + str2 != str2 + str1 {
        return ""
    } else {
        return str1[:gcd(len(str1), len(str2))]
    }
}